### PR TITLE
Revive tensorflow (C++) backend.

### PR DESCRIPTION
**Pull request by mooskagh**
Included fused batch support, a0 policy and WDL head.

It will be superseded by tensorflow C backend, but it's good to have it around for debugging.

### CODE REVIEW

**Comment by Tilps** [comment]
Do you have some check backend comparisons for the various networks to go with this?

**Comment by mooskagh** [comment]
Not sure how to use backend=check correctly but `./lc0 --backend=check --backend-opts=mode=histo,cudnn,tensorflow-cc` produces graphs like this:
```
Absolute error histogram for a batch of 178
 0.15 +                                                                                         +
      |                                                 ###                                     |
      |                                                 ###                                     |
      |                                                ####                                     |
      |                                                ####                                     |
  0.1 +                                               ######                                    +
      |                                               ######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
 0.05 +                                            ##########                                   +
      |                                            ##########                                   |
      |                                           ###########                                   |
      |#                                        # ############                                  |
      |#                               # #########################                              |
      +----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
   -inf  -15  -14  -13  -12  -11  -10   -9   -8   -7   -6   -5   -4   -3   -2   -1    0    1  +inf
```

**Comment by borg323** [comment]
This seems OK, but a safer comparison is against the blas backend.

**Comment by mooskagh** [comment]
It's been very long since I was able to build blas backend.
```
clang++  -o lc0 'lc0@exe/common_kernels.o' 'lc0@exe/fp16_kernels.o' 'lc0@exe/src_main.cc.o' 'lc0@exe/src_utils_protomessage.cc.o' 'lc0@exe/src_engine.cc.o' 'lc0@exe/src_version.cc.o' 'lc0@exe/src_benchmark_backendbench.cc.o' 'lc0@exe/src_benchmark_benchmark.cc.o' 'lc0@exe/src_chess_bitboard.cc.o' 'lc0@exe/src_chess_board.cc.o' 'lc0@exe/src_chess_position.cc.o' 'lc0@exe/src_chess_uciloop.cc.o' 'lc0@exe/src_mcts_node.cc.o' 'lc0@exe/src_mcts_params.cc.o' 'lc0@exe/src_mcts_search.cc.o' 'lc0@exe/src_mcts_stoppers_factory.cc.o' 'lc0@exe/src_mcts_stoppers_stoppers.cc.o' 'lc0@exe/src_mcts_stoppers_timemgr.cc.o' 'lc0@exe/src_neural_cache.cc.o' 'lc0@exe/src_neural_encoder.cc.o' 'lc0@exe/src_neural_factory.cc.o' 'lc0@exe/src_neural_loader.cc.o' 'lc0@exe/src_neural_network_check.cc.o' 'lc0@exe/src_neural_network_demux.cc.o' 'lc0@exe/src_neural_network_legacy.cc.o' 'lc0@exe/src_neural_network_mux.cc.o' 'lc0@exe/src_neural_network_random.cc.o' 'lc0@exe/src_neural_network_rr.cc.o' 'lc0@exe/src_neural_writer.cc.o' 'lc0@exe/src_selfplay_game.cc.o' 'lc0@exe/src_selfplay_loop.cc.o' 'lc0@exe/src_selfplay_tournament.cc.o' 'lc0@exe/src_syzygy_syzygy.cc.o' 'lc0@exe/src_utils_commandline.cc.o' 'lc0@exe/src_utils_configfile.cc.o' 'lc0@exe/src_utils_esc_codes.cc.o' 'lc0@exe/src_utils_histogram.cc.o' 'lc0@exe/src_utils_logging.cc.o' 'lc0@exe/src_utils_optionsdict.cc.o' 'lc0@exe/src_utils_optionsparser.cc.o' 'lc0@exe/src_utils_random.cc.o' 'lc0@exe/src_utils_string.cc.o' 'lc0@exe/src_utils_weights_adapter.cc.o' 'lc0@exe/src_utils_filesystem.posix.cc.o' 'lc0@exe/src_neural_network_tf_cc.cc.o' 'lc0@exe/src_neural_tensorflow_network_tf.cc.o' 'lc0@exe/src_neural_blas_convolution1.cc.o' 'lc0@exe/src_neural_blas_fully_connected_layer.cc.o' 'lc0@exe/src_neural_blas_se_unit.cc.o' 'lc0@exe/src_neural_blas_network_blas.cc.o' 'lc0@exe/src_neural_blas_winograd_convolution3.cc.o' 'lc0@exe/src_neural_opencl_network_opencl.cc.o' 'lc0@exe/src_neural_opencl_OpenCL.cc.o' 'lc0@exe/src_neural_opencl_OpenCLTuner.cc.o' 'lc0@exe/src_neural_opencl_OpenCLBuffers.cc.o' 'lc0@exe/src_neural_shared_activation.cc.o' 'lc0@exe/src_neural_shared_winograd_filter.cc.o' 'lc0@exe/src_neural_cuda_network_cudnn.cc.o' 'lc0@exe/src_neural_cuda_layers.cc.o' -Wl,--as-needed -Wl,--no-undefined -pthread -Wl,--start-group -ldl /usr/lib/libtensorflow_cc.so /usr/lib/libtensorflow_framework.so /usr/lib/libtensorflow.so /usr/lib/libopenblas.so /usr/lib/libOpenCL.so /opt/cuda/lib64/libcublas.so /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/9.2.1/../../../../lib64/libcudnn.so /opt/cuda/lib64/libcudart.so /usr/lib/libz.so /usr/local/lib/libprofiler.so -Wl,--end-group -Wl,-rpath,/opt/cuda/lib64:/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/9.2.1/../../../../lib64:/usr/local/lib -Wl,-rpath-link,/opt/cuda/lib64 -Wl,-rpath-link,/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/9.2.1/../../../../lib64 -Wl,-rpath-link,/usr/local/lib
/usr/bin/ld: lc0@exe/src_neural_blas_convolution1.cc.o: in function `lczero::Convolution1::Forward(unsigned long, unsigned long, unsigned long, float const*, float const*, float*)':
/home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/convolution1.cc:60: undefined reference to `cblas_sgemm'
/usr/bin/ld: lc0@exe/src_neural_blas_fully_connected_layer.cc.o: in function `lczero::FullyConnectedLayer::Forward1D(unsigned long, unsigned long, unsigned long, float const*, float const*, float const*, bool, float*)':
/home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/fully_connected_layer.cc:62: undefined reference to `cblas_sgemv'
/usr/bin/ld: /home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/fully_connected_layer.cc:93: undefined reference to `cblas_sgemm'
/usr/bin/ld: lc0@exe/src_neural_blas_fully_connected_layer.cc.o: in function `lczero::FullyConnectedLayer::Forward0D(unsigned long, float const*, float const*)':
/home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/fully_connected_layer.cc:138: undefined reference to `cblas_sdot'
/usr/bin/ld: lc0@exe/src_neural_blas_winograd_convolution3.cc.o: in function `lczero::WinogradConvolution3::Sgemm(unsigned long, float const*, unsigned long, unsigned long)':
/home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/winograd_convolution3.cc:222: undefined reference to `cblas_sgemm'
/usr/bin/ld: /home/crem/dev/lc0/build/debugoptimized/../../src/neural/blas/winograd_convolution3.cc:222: undefined reference to `cblas_sgemm'
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
```

I can compare with opencl if that helps:
```
Absolute error histogram for a batch of 222
 0.15 +                                                                                         +
      |                                                 ##                                      |
      |                                                 ###                                     |
      |                                                ####                                     |
      |                                                ####                                     |
  0.1 +                                               ######                                    +
      |                                               ######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
 0.05 +                                            ##########                                   +
      |                                            ##########                                   |
      |                                           ###########                                   |
      |                                         # ############                                  |
      |#                               # #########################                              |
      +----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
   -inf  -15  -14  -13  -12  -11  -10   -9   -8   -7   -6   -5   -4   -3   -2   -1    0    1  +inf
```

**Comment by mooskagh** [comment]
With BLAS/eigen
```
Absolute error histogram for a batch of 256
      |                                                                                         |
 0.15 +                                                  #                                      +
      |                                                 ##                                      |
      |                                                 ###                                     |
      |                                                ####                                     |
      |                                                ####                                     |
  0.1 +                                               #####                                     +
      |                                               ######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
      |                                              #######                                    |
 0.05 +                                            ##########                                   +
      |                                            ##########                                   |
      |                                           ###########                                   |
      |                                         # ############                                  |
      |#                               ##########################                               |
      +----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
   -inf  -15  -14  -13  -12  -11  -10   -9   -8   -7   -6   -5   -4   -3   -2   -1    0    1  +inf
```

File: src/neural/network_tf_cc.cc [hunk 1]
```diff
@@ -132,18 +138,49 @@ std::pair<Output, Output> MakeNetwork(const Scope& scope, Input input,
...
+    auto conv_pol =
+        MakeConvBlock<CPU>(scope, conv_pol1, 3, filters, 80, weights.policy,
+                           nullptr, nullptr, /* relu= */ false);
+
+    // [1858 -> HWC or CHW]
+    std::vector<int> policy_map(1858);
+    for (const auto& mapping : kConvPolicyMap) {
+      if (mapping == -1) continue;
+      const auto index = &mapping - kConvPolicyMap;
+      const auto direction = index / 64;
```
Comment (Tilps) [inline src/neural/network_tf_cc.cc]: 'direction' doesn't seem the most obvious name.  Its 73 options for different moves relative to square, for which there are only 16 distinct 'directions'.

Comment (mooskagh) [reply src/neural/network_tf_cc.cc]: Changed to "displacement".

Comment (Tilps) [reply src/neural/network_tf_cc.cc]: better - we can think of it as a 3 dimensional displacement, 2d of movement and one of piece type change :P

**Pull request merged**